Migrate Azure settings to new settings infrastructure

danielmitterdorfer opened the pull request:
With this commit we migrate all Azure related settings to
the new settings infrastructure.

During the implementation I have spotted a few things which I have improved along the way:

* Made some classes immutable
* Removed usage of outdated classes like `StringBuffer` and `Hashtable`
* Removed leniency in `AzureComputeServiceImpl` which allowed any garbled value as key store type and just assumed some default
* Fixed some tests that have apparently not been executed in CI

Can you have a look at it @s1monw?

dadoonet commented on plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/AzureDiscoveryModule.java:
```diff
@@ -79,29 +80,24 @@ public static boolean isDiscoveryReady(Settings settings, ESLogger logger) {
-                isPropertyMissing(settings, Management.KEYSTORE_PASSWORD)
-            ) {
-            logger.debug("one or more azure discovery settings are missing. " +
+        if (isDefined(settings, Management.SUBSCRIPTION_ID_SETTING) &&
```
Definitely better

s1monw commented:
this looks awesome!
LGTM thank you so much

dadoonet commented on plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeService.java:
```diff
@@ -19,31 +19,25 @@
-        public static final String KEYSTORE_TYPE = "cloud.azure.management.keystore.type";
+        public static final Setting<String> KEYSTORE_PATH_SETTING = Setting.simpleString("cloud.azure.management.keystore.path", false, Setting.Scope.CLUSTER);
+        public static final Setting<String> KEYSTORE_PASSWORD_SETTING = Setting.simpleString("cloud.azure.management.keystore.password", false, Setting.Scope.CLUSTER);
+        public static final Setting<KeyStoreType> KEYSTORE_TYPE_SETTING = new Setting<>("cloud.azure.management.keystore.type", KeyStoreType.pkcs12.name(), KeyStoreType::fromString, false, Setting.Scope.CLUSTER);
```
Just for my understanding: what happens if the user provides a wrong type? Does it fail starting the node or does it fallback to `KeyStoreType.pkcs12`?

danielmitterdorfer replied:
If nothing is provided then it will fall back on the default value. If an invalid value is provided, it will fail.

dadoonet replied:
Thanks @danielmitterdorfer. So I think we should modify https://github.com/elastic/elasticsearch/pull/16291/files#diff-3d1a858d84f619a449e4a1d8829fc69eR91 and fails instead of fallbacking so we are consistent? WDYT?

danielmitterdorfer replied:
I originally thought it would be ok to limit the chunk size to the maximum value but you're right. This would be even more explicit for users when we fail here. I'll change it accordingly.

dadoonet commented on plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureRepository.java:
```diff
@@ -60,31 +60,23 @@
+        public static final Setting<String> CONTAINER_SETTING = new Setting<>("container", "elasticsearch-snapshots", Function.identity(), false, Setting.Scope.CLUSTER);
+        public static final Setting<String> BASE_PATH_SETTING = Setting.simpleString("base_path", false, Setting.Scope.CLUSTER);
+        public static final Setting<String> LOCATION_MODE_SETTING = Setting.simpleString("location_mode", false, Setting.Scope.CLUSTER);
+        public static final Setting<ByteSizeValue> CHUNK_SIZE_SETTING = Setting.byteSizeSetting("chunk_size", new ByteSizeValue(64, ByteSizeUnit.MB), false, Setting.Scope.CLUSTER);
```
May be we could use here `MAX_CHUNK_SIZE` as the default value?

danielmitterdorfer replied:
Ah, I thought they were unrelated. That's why I kept them separate. I can change that. Thanks for clarifying.

danielmitterdorfer commented:
Thanks for the feedback. I'll look whether I can migrate `AzureStorageSettings` similarly to [`EsLoggerFactory`](https://github.com/elastic/elasticsearch/pull/16289/files#diff-7117bca712d8a68a5aa84783e75d902aR40) though.

dadoonet commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureSnapshotRestoreTests.java:
```diff
@@ -106,9 +106,9 @@ public void testSimpleWorkflow() {
-                        .put(Repository.CONTAINER, getContainerName())
-                        .put(Repository.BASE_PATH, getRepositoryPath())
-                        .put(Repository.CHUNK_SIZE, randomIntBetween(1000, 10000), ByteSizeUnit.BYTES)
+                        .put(Repository.CONTAINER_SETTING, getContainerName())
```
Shouldn't this be `Repository.CONTAINER_SETTING.getKey()`?

danielmitterdorfer replied:
It should be. I'm puzzled why this didn't fail in my local verification build. Maybe because it's a third party test case? I'll correct the key anyway.

danielmitterdorfer commented:
I've integrated the review comments by @dadoonet now. 

Apart from that, I found the implementation in `AzureStorageSettings#parse()` very hard to understand. Here's what the method basically does:

Suppose we have the following properties defined in the config file (please ignore that some properties are missing):

```
cloud.azure.storage.azure1.timeout=30s
cloud.azure.storage.azure1.account=demo
cloud.azure.storage.azure1.default=true

cloud.azure.storage.azure2.timeout=45s
cloud.azure.storage.azure2.account=demo2

cloud.azure.storage.azure3.timeout=45s
cloud.azure.storage.azure3.account=demo3
```

For each "group" ("azure1", "azure2", "azure3"), it will create an instance of `AzureStorageSettings`, verify a few constraints (e.g. there must be only one group with `default` set to `true`) and return the "primary" (which is the group with `default` set to `true`) and a list of secondaries (which are all other groups).

I have added an experimental reimplementation of this method based on the following approach:

* Partition the provided settings by group key (will create three groups "azure1", "azure2" and "azure3" considering the example above). The respective setting keys will be "normalized" by group. So instead of `cloud.azure.storage.azure3.timeout`, the setting key will be `cloud.azure.storage.timeout` internally. This allows us to use the settings infrastructure here without creating `Setting` instances for each group.
* Based on these settings, it creates preliminary `AzureStorageSettings`, one for each group
* It will validate global configuration rules not on raw config values but based on the `AzureStorageSettings` instances created before. It is also now more restrictive, in the sense that it will fail if multiple groups have `default` set to `true`.

For more details, please see also my commit comment in 0d88586ef8a8d8174637b9eb2cb5cdc95e695acd. I have implemented the experimental parsing logic in the separate commit 0d88586ef8a8d8174637b9eb2cb5cdc95e695acd, so it is easy to revert if the implementation does not make too much sense.

I'd like to get your feedback (especially from @dadoonet as you know a lot about this module and @s1monw as you've written the settings infrastructure). If this doesn't make sense for you then I'm happy to revert the commit and merge the PR without this change but I wanted to give it at least a try.

Advantages from my point of view:

* The code is easier to understand (at least from my point of view) although it is now more code
* It is stricter and rules are explicitly checked on domain objects instead of implicit checks during parsing
* We can leverage the settings infrastructure

Disadvantages:
* We introduce "synthetic" setting keys which could mislead users on configuration errors :(
* I'd prefer if we could integrate this better into the settings infrastructure but I didn't see how this could fit in, even though we have now `Setting#dynamicKeySetting()` as we'd need to be able to group multiple values by infix.

Open issues:
* Tests for the new code (I didn't want to burn too much time and just wanted to get early feedback from you before proceeding)

danielmitterdorfer commented:
I've merged the PR now without the refactored parse method (see my previous comment) and will do this in a separate PR.
